### ISSUE TITLE
Change the Unit and VSSDataType to support custom types and units

### ISSUE THREAD

**Issue by oliveirarleo** [GENIVI/vss-tools#76]
In Python we are note able to create a subclass from an Enum, so we have no alternative to the provided _unit_  and _data\_type_ from the **Units** and **VSSDataTypes** classes. So in parse-time these classes are used to type-check and it's not possible to use any type or unit that is not available as standard in this repository.

In my project we are using more _unit_  and _data\_type_  than the provided ones. We would like to be able to have a way to add new types in runtime for both of these classes.

To solve that we suggest a registry pattern in a way that aside from the standard types, we can add more attributes to the class by using a method **register(name, value)** for example. 

With that said, int the case of the **Unit** class it could be interesting to add the following units as standard in your library as well. 
```
    LUX="lux"
    HERTZ="Hz"
    MILLISECOND="ms"
```

**Comment by Floix** [comment]
"So in parse-time these classes are used to type-check and it's not possible to use any type or unit that is not available as standard in this repository." This is actually intended behavior, in that sense that adding Units requires a Pull Request. In the past we had multiple instances of liters (l, liter, liters) and no overview of the units that were used and supported, which again makes tool development very hard because you do not know what your unit API is.

If you add those using a registry you'll end up with the same problem. I would recommend opening a Pull Request to add units that you require and make it part of the standard.

**Comment by SebastianSchildt** [comment]
I would prefer a solution where the tools can read a list of allowed units (and provide a default "VSS standard catalgoue" unit file).

The problem ist, "lux" is probably a nice addition, but I would refrain from adding it to the VSS (standard catalogue) as long as no standard entry uses it. Delegating this in a _configurable_ way to the tools the same way you can also convert other "Non-standard" VSS trees with it), seems to me the desirable solution.

**Comment by alexanderdomin** [comment]
I also like the idea having a core of supported types and units, plus giving the users the possibility to extend units and types as required. This approach will seed up the development process and boost the tool acceptance. I think some of the people do not have much time to discuss if a new type or unit shall be added to the standard and how to name the new type or unit.

**Comment by Floix** [comment]
I think this is difficult proposal. The whole purpose of VSS would be to write down a specification of how we agree on components of a vehicle. Part of the core spec are units and the tool supports and validates that. If we go down the road and allow adding basically any unit during runtime, our spec is becoming less reliable. Ultimatly we do not end up with a model, but with a rough idea that is not validated nor really specified. 

We are talking here about 3 lines of spec to add 3 units here: https://github.com/GENIVI/vss-tools/blob/14d044699bc358590009aaf1740864c149d3c9be/vspec/model/constants.py#L83. If you do not have the time to propose that, you certainly won't find the time to do all the validation in your code that you now need to implement as the tool validation is less reliable.

What you could do is, add a flag to turn of all validation, if that helps your code. For any sort of "Proof of Concept" tooling that's propably enough.

**Comment by barbieri** [comment]
for us it's more a matter of a decision, we had a fork with the units added, indeed much easier.

However the extensible registry is more a matter of PoC and experimentation as you pointed out.

BTW, what about the datatypes? There is a similar issue, for instance we'll need a color... we went with `HSVColor` (on our side we're handing the type in transport as a string encoded as `"hsv({hue}, {saturation}%, {value}%)"`), but `RGBColor` would work. We also would like to have some more complete `DateTime` with timezone included (instead of UNIX Epoch, something like ISO8601).

**Comment by gunnarx** [comment]
Add your own data types?   Spontaneously, no.  @Floix explains why.  We are aiming to make a common and everywhere compatible "model" for how data can be described.  It is not desirable to have incompatible data models, so that only some programs will understand the data types that are used, and others will see it as nonsense.   It is not a goal to create fragmentation in the basic model itself.   It is possible that some _data items_ are "private", but the rules for how these are represented should be the same for all, public and private, catalogs of data.  We should hold strong on that point.   First, please, clarify what data type you are missing, and why?

For units, I'd be a bit more open, with either @SebastianSchildt or @Floix solution, but I'm leaning towards @Floix.  If you could explain first which types of signals you want to add, and which units you are missing for those?  Maybe it's something useful to the standard VSS.

Let's talk about why you are proposing this.  If it is for vehicle signals (realm of VSS) I'd say the idea is that you give your input here to affect the _standard_ which is then used by everyone.   

If you are rather thinking of using this for a kind of non-vehicle approach, which might have different needs, then we should open up the larger discussion (again), which is to set a name to the "VSS-like" data modeling approach usable for multiple purposes, where VSS is a subset/instance of that more generic data modeling concept.  (We already have others that are sort of asking for that possibility).  In alternative instances of the high-level generic data model then alternative datatypes might be allowed.  (But again, which are they, beyond numbers and strings (and stream URLs))?

I would also like to make a note of another standard answer, if it helps here, and that is to consider working with us on the [VSC / Generic IDL work](https://github.com/GENIVI/vehicle_service_catalog) (not too much at that link except an example, but there is more information if you are interested).  Because that is going in the direction of a fully flexible and modeling language (define "any" data type, and so on).  In other words, maybe some needs could be met by using that instead.

**Comment by gunnarx** [comment]
Continuing to answer @barbieri 
> for us it's more a matter of a decision, we had a fork with the units added, indeed much easier.

Works, for short time.  We encourage designing VSS for the needs of automotive companies instead of having forks internally.  (Have you proposed the changes upstream yet?)

> However the extensible registry is more a matter of PoC and experimentation as you pointed out.

Not a big need, as I see it.  For any type of experimentation, local fork is the standard way.  It would encourage that it does not stay forked forever better than making it configurable, IMHO.

> BTW, what about the datatypes? There is a similar issue, for instance we'll need a color... we went with HSVColor (on our side we're handing the type in transport as a string encoded as "hsv({hue}, {saturation}%, {value}%)"), but RGBColor would work. 

I've heard about this one.   If we need to bring up this discussion again (hopefully to put it to rest), let's do it.  As far as I understand, fixed-size arrays are now introduced and can handle this in a rudimentary way.  (To simply define a fixed-size tuple is why I voted for uint32[3] syntax instead of uint32[] and defining the size separately, but was voted down).

If you want strictly typed tuples... I'm not sure that's the VSS way really.  I think we are then in full-featured VSC/IDL realm, see above.

> We also would like to have some more complete DateTime with timezone included (instead of UNIX Epoch, something like ISO8601).


Separate ticket or PR, please.

**Comment by barbieri** [comment]
> Continuing to answer @barbieri
> 
> > for us it's more a matter of a decision, we had a fork with the units added, indeed much easier.
> 
> Works, for short time. We encourage designing VSS for the needs of automotive companies instead of having forks internally. (Have you proposed the changes upstream yet?)

we're opening this (and other tickets) in the process of us doing exactly that. So for units and datatypes, we'll provide the patch to VSS adding more elements (instead of a registry), as per this discussion :-)


> > However the extensible registry is more a matter of PoC and experimentation as you pointed out.
> 
> Not a big need, as I see it. For any type of experimentation, local fork is the standard way. It would encourage that it does not stay forked forever better than making it configurable, IMHO.

fair enough


> > BTW, what about the datatypes? There is a similar issue, for instance we'll need a color... we went with HSVColor (on our side we're handing the type in transport as a string encoded as "hsv({hue}, {saturation}%, {value}%)"), but RGBColor would work.
> 
> I've heard about this one. If we need to bring up this discussion again (hopefully to put it to rest), let's do it. As far as I understand, fixed-size arrays are now introduced and can handle this in a rudimentary way. (To simply define a fixed-size tuple is why I voted for uint32[3] syntax instead of uint32[] and defining the size separately, but was voted down).
> 
> If you want strictly typed tuples... I'm not sure that's the VSS way really. I think we are then in full-featured VSC/IDL realm, see above.

I think handling as array is not the way to go, for rgb okay, they are all uint8_t, so 3 bytes do it. But for HSV we already have something different (we could fallback to the largest data type, such as 3 doubles), but being explicit will help (as we have boolean, instead of uint8_t with min=0, max=1)

By presenting a higher level scope with this "scalar" (the GraphQL term), we can do better mapping to various techs, like we could use native Android/iOS types, we could offer well defined conversions in the user (like RGB<->HSV).

I see below you point the DateTime to be handled as a different ticket (will do), but the idea is also the same, by stating it's date, time and timezone, we can clearly communicate and use it for such purpose, while we could use Unix Epoch for others (like when timezones are not changing or irrelevant). Keeping them in the same "scalar" (again, using GraphQL term) helps us keep it consistent and unified (if we try to carry another field "timezone", it's easy to get wrong at various layers).

 
> > We also would like to have some more complete DateTime with timezone included (instead of UNIX Epoch, something like ISO8601).
> 
> Separate ticket or PR, please. (And both model/rules and content is defined in https://github.com/genivi/vehicle_signal_specification so that's probably where the proposal should go)

ok

**Comment by alexanderdomin** [comment]
@gunnarx it was nice to read the thread and i my comments I would like to focus only on the units:

1. I do not understand why we can not have this flexibility :-/. I am sure people working with VSS will patch and add own Units. And you can not prevent this, its a business decision. With this concept we can make live easy for the developers.
2. If we take Franca as an example, the unit is a part of documentation and will be not enforced by the tooling. Client developer is responsible to read the documentation and do proper conversion. ´By the way android is also using units as comment in the documentation.
3. In VSS we try to be moor strict. But I do not understand one thing. On the this page https://genivi.github.io/vehicle_signal_specification/rule_set/data_entry/data_unit_types/ I can find "ms" and "Hz" but in this python script https://github.com/GENIVI/vss-tools/blob/1763af749c1110fb090a79bfc749f684c663e6b2/vspec/model/constants.py I can not find this units. So before we can start discussing missing units I propose you add all missing units from existing documentation to the python script.
4. And to shorten the discussion on the units I think its will save our time if we add all possible units to the documentation and the python script. Sooner or later someone will need it and we do not have to discuss and developers do not have to fork.

**Comment by Floix** [comment]
"With this concept we can make live easy for the developers." and "Client developer is responsible to read the documentation and do proper conversion." are contradicting each other. If you have a clear and sound contract of the data you are expecting you make live easier for the developers unless you are experimenting with stuff. Once you move towards production you make your life extremly hard when you are flexible on the data model. If we want VSS to be a model for prototyping some stuff than flexibility is the right decisions. If we want to develop VSS into something that is relevant in productive use cases than flexibility will be a money/capacity drain in the mid- to long-run.

The current code is based on the units there were already added when we added unit checking. The documentation was not on sync with the model at that time and propabily still isn't 100%. That's exatcly the reason why we added checking. A wiki page would not enforce rules on the model. However, I think it's a good idea to try to keep those list in sync. Would you open a PR to add the missing units @alexanderdomin ?

I think it's not the best idea to add all possible units. You'll have to react on all of these units in your application even if you do not use them. Think about bar and kpa as an example. If you limit yourself to use one of those you'll never have to do any checks and fallback implementations. Worst case scenario is a conversion for your frontend for instance. That's one of the reaons we should limit ourselves to SI units as proposed here: https://genivi.github.io/vehicle_signal_specification/rule_set/data_entry/data_unit_types/


I currently see only the following options:
1.) You need a new unit ==> Open a PR and add it. Discussion about adding a certain unit will improve the overall quality of the data model and keep it clean and simple.

2a.) You experiment with VSS and want to check some units: Fork and test, once you completed your experiment and want  to move on using VSS propose a change to the units and continue with 1.)

2b.) Add an option to deactive type checking for your own use. Similar to 2, but instead of a fork you call the tools with a certain parameter.  Once you completed your experiment and want  to move on using VSS propose a change to the units and continue with 1.)

**Comment by alexanderdomin** [comment]
> "With this concept we can make live easy for the developers." and "Client developer is responsible to read the documentation and do proper conversion." are contradicting each other. If you have a clear and sound contract of the data you are expecting you make live easier for the developers unless you are experimenting with stuff. Once you move towards production you make your life extremly hard when you are flexible on the data model. If we want VSS to be a model for prototyping some stuff than flexibility is the right decisions. If we want to develop VSS into something that is relevant in productive use cases than flexibility will be a money/capacity drain in the mid- to long-run.

And this is my point: as soon as we go in production we will fork and extend units as we need until we added everything required. And we will have not time and resource to cover internal OSS upstream processes and endless upstream discussion. This is reality. Goolge Vehicle Properties concept is more flexible and I would like to see such flexibility in our tooling.

> 
> The current code is based on the units there were already added when we added unit checking. The documentation was not on sync with the model at that time and propabily still isn't 100%. That's exatcly the reason why we added checking. A wiki page would not enforce rules on the model. However, I think it's a good idea to try to keep those list in sync. Would you open a PR to add the missing units @alexanderdomin ?

I think it is much faster if the @danielwilms creates this PR. I need to write several OSS contribution documents and approve by several department leads and software developers. 
 
> 
> I think it's not the best idea to add all possible units. You'll have to react on all of these units in your application even if you do not use them. Think about bar and kpa as an example. If you limit yourself to use one of those you'll never have to do any checks and fallback implementations. Worst case scenario is a conversion for your frontend for instance. That's one of the reaons we should limit ourselves to SI units as proposed here: https://genivi.github.io/vehicle_signal_specification/rule_set/data_entry/data_unit_types/

I am talking about grammar definition. I am not talking about providing several units for the same leaf. So if you take tire pressure we defined it in bar. But you can have other pressure leafs in you vehicle and this can be expressed also via kpa.

**Comment by barbieri** [comment]
@Floix it makes sense... but to fix the docs x practice, it would be nice to have some tooling to keep them in sync, like:
 - generate the JSON schema from the Python enum
 - generate a `.md` (markdown) from the Python enum

this way they're always in sync.

That said, we'll proceed with the patches... I know @alexanderdomin's pain, I'd not mind the flexibility of the registry, but I also see the effort to make it consistent, like a single unit for all speed (not km/h, mi/h...) and such.

We can do this outside of the paid contract, so no worries about the paperwork @alexanderdomin, it's a super small task, that one comes for free ;-)

**Comment by barbieri** [comment]
@Floix added #81 that generates markdown from the code... it's a suggestion on how we could generate the doc, keeping it in sync (but many cases are still missing, I marked some missing descriptions/domains as `?`).

**Comment by Floix** [comment]
I had a brief check on the PR. I'm not sure if it serves the purpose. Instead of simplifying the approach, we now added additional requirements to introduce units?

**Comment by danielwilms** [comment]
@oliveirarleo: will create PR and link
